sockeye/encoder.py

```diff
@@ -309,6 +326,66 @@ def get_num_hidden(self) -> int:
         return self.config.num_embed
 
 
+class FactoredEmbedding(Embedding):
+    """
+    Thin wrapper around MXNet's Embedding symbol. Works with both time- and batch-major data layouts.
+    Assumes the main word embedding weights are passed in, allowed them to be shared with the target.
+    Builds the embeddings for the source factors.
+
+    :param config: Embedding config.
+    :param prefix: Name prefix for symbols of this encoder.
+    :param factor_vocab_sizes: List of source factor vocabulary sizes.
+    :param factor_output_dims: List of source factor output dimensions.
+    :param embed_weight: Optionally use an existing embedding matrix instead of creating a new one.
+    """
+
+    def __init__(self,
+                 config: FactoredEmbeddingConfig,
+                 prefix: str,
+                 factor_vocab_sizes: List[int],
+                 factor_output_dims: List[int],
+                 embed_weight: Optional[mx.sym.Symbol] = None) -> None:
+        super().__init__(config, prefix, embed_weight)
+
+        self.factor_vocab_sizes = factor_vocab_sizes
+        self.factor_output_dims = factor_output_dims
+        self.num_factors = len(self.factor_vocab_sizes)
+
+        # Factors weights aren't shared so they're not passed in and we create them here.
+        self.embed_factor_weights = [
+            mx.sym.Variable(prefix + "factor" + str(i) +  "_weight",
+                            shape=(self.factor_vocab_sizes[i],
+                                   self.factor_output_dims[i]))
+            for i in range(self.num_factors)]
+
+    def encode(self,
+               data: mx.sym.Symbol,
+               data_length: mx.sym.Symbol,
+               seq_len: int,
+               factors: List[mx.sym.Symbol]) -> Tuple[mx.sym.Symbol, mx.sym.Symbol, int]:
+        """
+        Encodes data given sequence lengths of individual examples and maximum sequence length.
+
+        :param data: Input data.
+        :param data_length: Vector with sequence lengths.
+        :param seq_len: Maximum sequence length.
+        :param factors: The factors to encode.
+        :return: Encoded versions of input data (data, data_length, seq_len).
+        """
+
+        embeddings = [ super().encode(data, data_length, seq_len)[0] ]
```

Comment:
I think it would be nicer to get the length return from the word embeddings and return that, something like:
```
word_embeddings, embed_length, embed_seq_len = super.encode(data, data_length, seq_len)
factor_embeddings = []
[...]
embedding = mx.sym.concat(word_embeddings, factor_embeddings, ...)
return embedding, embed_length, embed_seq_len
```